[ENH] Generalize splitters to accept timedeltas (equally spaced)

khrapovs opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
https://github.com/alan-turing-institute/sktime/issues/1737

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->

- All splitters now accept not only integer, but also `timedelta` or `pd.DateOffset` for `initial_window`, `window_length`, and `step_length` arguments. No regression was introduced. The core changes are localized to `sktime/forecasting/model_selection/_split.py`.
- All existing tests (`sktime/forecasting/model_selection/tests/test_split.py`) are only enriched for additional argument types. Test cases boil down to adding new types in `sktime/forecasting/tests/_config.py`. See variables `TEST_WINDOW_LENGTHS`, `TEST_INITIAL_WINDOW`, and `TEST_STEP_LENGTHS`.
- I could not resist and added type hints in `sktime/forecasting/base/_fh.py` although it is not necessary for the core change. I added type hints to `_split.py` first, but for completeness I needed type hint for forecasting horizon which triggered the change in `_fh.py`.
- `sktime/forecasting/model_evaluation/tests/test_evaluate.py` was only changed to preserve original tests. Other changes are automatic commit hooks. Same for `sktime/forecasting/model_selection/tests/test_tune.py`, `sktime/forecasting/tests/test_naive.py`, and `sktime/forecasting/tests/test_all_forecasters.py`.
- In `sktime/utils/validation/__init__.py` and `sktime/utils/validation/forecasting.py` I have added a few validations for new types.

#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

none

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

The focus of this PR are the new types of above mentioned arguments. There are relatively a lot of changes, but all of them are triggered by this generalization.

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

This is the first, but hopefully, logically complete batch of changes. I have a much bigger target in mind (see #1737). Personally, I prefer smaller PRs, hence I would prefer to stop here and address other required changes (like generalization of `ForecastHorizon`) in another PR.

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.

<!--
Thanks for contributing!
-->

fkiraly commented:
hello @khrapovs, how is it going? Do you need any help or feedback with this?

khrapovs commented:
Hello, Franz. Happy New Year! I decided to switch off for one week while I am on holidays with my family. Will come back to this issue upcoming week. It goes very well, I do not see any major obstacles. Just need a bit time to generalize a few functions. Eventually I would love to have it scrutinized by someone experienced, but right now my changes are too dirty.

khrapovs commented:
@fkiraly I think it is time to show this to someone. Of course, it does not implement everything I need for my use case, but this is already logically complete and I would love to proceed with a review here.

fkiraly reviewed (commented):
ok, quick question, would you be able to give a three-bullet-point summary of the changes you made so far?
(i.e., which functions, modules, classes are changed how?)

Is it correct that the changes are localized to `_split` and tests, plus adding type hints in every file you saw?

khrapovs commented:
@fkiraly I have updated the description of my PR at the top as you requested. I hope it is sufficiently clear.

fkiraly commented:
thanks, will probably have time on Fri or Mon

fkiraly reviewed (changes requested):
Excellent work! We've been too closely stuck with the integer indexing, so using timedeltas will be useful for irregularly spaced series!

Looks good and logical on detail inspection.
However, I think it is crucial for review to make clear what the behaviour of the various splitters is in "timedelta" mode - the devil is in the detail.

Would you hence be able to add the following to the docs:
* proper docstrings for key methods, such as `split`, `get_n_splits` and `_split`. They were missing before (not your fault), but now with timedeltas it's crucial to explain well what happens.
* for all splitters, describe in the class docstring what kind of splits these generate, especially if you have the general setting.

Further: would it be easy to make the code robust with respect to unequally sampled time series? Or difficult?

A minor question on the side:
We're looking at extending the forecasting functionality to panel and hierarchical time series. There, we would use `pd.DataFrame` with a `MultiIndex`, with time being the last level. May be better to extend that later, but do you foresee any substantial problems?

fkiraly commented on sktime/forecasting/base/_fh.py:
```diff
@@ -149,7 +155,11 @@ def __new__(cls, values=None, is_relative=None):
-    def __init__(self, values=None, is_relative=True):
+    def __init__(
+        self,
+        values: Optional[Union[VALID_FORECASTING_HORIZON_TYPES]] = None,
```
I think you don´t need the `Optional` if the default is already `None` (then `Optional` is inferred)?

khrapovs replied:
you are right. I have reread one more time https://docs.python.org/3/library/typing.html#typing.Optional and removed `Optional`

fkiraly commented on sktime/forecasting/model_evaluation/tests/test_evaluate.py:
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3 -u
 # -*- coding: utf-8 -*-
 # copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""Testt evaluate."""
```
Maybe a bit more descriptive? "Tests for model evaluation module", etc

khrapovs replied:
Now it reads:
```
"""Tests for model evaluation module.

In particular, function `evaluate`, that performs time series
cross-validation, is tested with various configurations for correct output.
"""
```

fkiraly replied:
ok, thanks!

fkiraly commented on sktime/forecasting/model_selection/_split.py:
```diff
@@ -158,42 +175,100 @@ def _get_end(y, fh):
+    if is_timedelta_or_date_offset(x=window_length):
+        if y.get_loc(min(y[-1], y[0] + window_length)) + fh_max > n_timepoints:
+            raise ValueError(
+                f"The `window_length` and the forecasting horizon are incompatible "
```
Here, I think it would be good to spell out the precise condition in the error message (which is violated)

khrapovs replied:
Added these two lines:
```
f"It is required that the window length plus maximum forecast horizon "
f"is smaller than the length of the time series `y` itself."
```
I have also extracted common error text into variables which are used twice.

fkiraly replied:
yes, ok. May be improved by ensuring this is clearly linked to functino/method inputs, but good enough.

fkiraly commented on sktime/forecasting/model_selection/_split.py:
```diff
@@ -158,42 +175,100 @@ def _get_end(y, fh):
-                f"The `initial_window` and the forecasting horizon are incompatible "
-                f"with the length of `y`. Found `initial_window`={initial_window},"
+                f"The `window_length` and the forecasting horizon are incompatible "
+                f"with the length of `y`. Found `window_length`={window_length}, "
```
same here

khrapovs replied:
same here

fkiraly commented on sktime/forecasting/model_selection/_split.py:
```diff
@@ -158,42 +175,100 @@ def _get_end(y, fh):
+                raise ValueError(
+                    f"The `initial_window` and the forecasting horizon are "
+                    f"incompatible "
+                    f"with the length of `y`. Found `initial_window`={initial_window}, "
```
same, and below

khrapovs replied:
Slightly changed wording:
```
"The `initial_window` and `window_length` types are incompatible. "
"They should be either all timedelta or all int."
```

fkiraly reviewed (changes requested):
Thanks for the docstrings!

I think these are currently major issue with the `_split` module, they need to become clearer.

I'd kindly request: in `BaseSplitter`, explain the type of object that this models - a collection of training and test splits. Explain which methods can be used to retrieve them.

In the classes that inherit from `BaseSplitter`, explain in addition what the split precisely will be,  depending on the parameters.
This should not just be examples, but a clear, explicit statement of what the splits are. Optimally, use mathematical or code formalism to give the description.

Otherwise, the changes look ok - but I'd give it another go and compare actual to intention, once you've written clear docstrings that describe what splitters you are *intending* to implement (so I can validate the timedelta case).

khrapovs commented:
@fkiraly As requested, I have introduced math formalism in the main section of the `BaseSplitter`. In order to avoid duplication I have referenced its docstring in concrete implementations and added missing details relative to the base. Please, let me know if it is still does not sound sufficiently clear in terms of intentions. Thanks again for a very detailed review!

fkiraly reviewed (approved):
@khrapovs, thanks! The changes to the docstring are quite good!
Good for approval, since the requirement was to understand what you were trying to do.

Couple of things could be improved, I leave this to the next PRs:
* explanation is not precise or a bit vagout about that a splitter encodes *multiple splits*, i.e., there is a collection of k, ell. If notation is introduced in the `BaseSplitter` for multiple k, ell, this can be referred to in concrete implementation
* explanation is very verbose, could be shortened
* explanation of methods and variables is a bit scattered, could be more list-like and concise
* you probably want to pass to a framing in terms of intervals, i.e., start/end of the window. In the end state, your observation time stamps may not need to be exactly at the start and end of a window that the splitter parameterizes
